Realm/RLMAccessor.hpp

```diff
@@ -0,0 +1,116 @@
+    id operator*() const noexcept { return value; }
+};
+
+class RLMAccessorContext {
+public:
+    // Accessor context interfact
+    RLMAccessorContext(RLMAccessorContext& parent, realm::Property const& property);
+
+    id box(realm::List&&);
+    id box(realm::Results&&);
+    id box(realm::Object&&);
+    id box(realm::RowExpr);
+
+    id box(realm::BinaryData v) { return RLMBinaryDataToNSData(v); }
+    id box(bool v) { return @(v); }
+    id box(double v) { return @(v); }
+    id box(float v) { return @(v); }
+    id box(long long v) { return @(v); }
+    id box(realm::StringData v) { return RLMStringDataToNSString(v); }
+    id box(realm::Timestamp v) { return RLMTimestampToNSDate(v); }
+    id box(realm::Mixed v) { return RLMMixedToObjc(v); }
+
+    id box(realm::util::Optional<bool> v) { return v ? @(*v) : nil; }
+    id box(realm::util::Optional<double> v) { return v ? @(*v) : nil; }
+    id box(realm::util::Optional<float> v) { return v ? @(*v) : nil; }
+    id box(realm::util::Optional<int64_t> v) { return v ? @(*v) : nil; }
+
+    void will_change(realm::Row const&, realm::Property const&);
+    void will_change(realm::Object& obj, realm::Property const& prop) { will_change(obj.row(), prop); }
+    void did_change();
+
+    RLMOptionalId value_for_property(id dict, std::string const&, size_t prop_index);
+    RLMOptionalId default_value_for_property(realm::ObjectSchema const&,
+                                             std::string const& prop);
+
+    template<typename Func>
+    void enumerate_list(__unsafe_unretained const id v, Func&& func) {
+        for (id value in v) {
+            func(value);
+        }
+    }
+
+    template<typename T>
+    T unbox(id v, bool create = false, bool update = false);
+
+    bool is_null(id v) { return v == NSNull.null; }
+    id null_value() { return NSNull.null; }
+    id no_value() { return nil; }
+    bool allow_missing(id v) { return [v isKindOfClass:[NSArray class]]; }
+
+    std::string print(id obj) { return [obj description].UTF8String; }
+
+    // Internal API
+    RLMAccessorContext(RLMObjectBase *parentObject, const realm::Property *property = nullptr);
+    RLMAccessorContext(RLMRealm *realm, RLMClassInfo& info, bool promote=true);
+
+    // The property currently being accessed; needed for KVO things for boxing
+    // List and Results
+    RLMProperty *currentProperty;
+
+private:
+    __unsafe_unretained RLMRealm *const _realm;
+    RLMClassInfo& _info;
+    // If true, promote unmanaged RLMObjects passed to box() with create=true
+    // rather than copying them
+    bool _promote_existing = true;
+    // Parent object of the thing currently being processed, for KVO purposes
+    __unsafe_unretained RLMObjectBase *const _parentObject = nil;
+
+    // Cached default values dictionary to avoid having to call the class method
+    // for every property
+    NSDictionary *_defaultValues;
+
+    bool _nilHack = false;
```

Comment:
right, comment probably not necessary then given that we'll be addressing that shortly.